shm_wrapper/README.md

```diff
@@ -20,12 +13,59 @@ make test2
+
+There are various methods of IPC; in order of (generally) slowest to fastest: file sharing, sockets, pipes and message queues, and shared memory. Since so much of the data is being transferred over IPC, and previous iterations of runtime using pipes still had problems with the speed of the IPC, the decision was made to use shared memory for c-runtime's IPC. Additionally, shared memory is well-suited for Runtime since the data that is passed between the processes is very structured and consistent (device data, gamepad data, robot state), so we don't have to worry about constantly expanding or shrinking the size of the shared memory blocks.
+
+### POSIX Shared Memory Overview
+
+We use POSIX shared memory (as opposed to the older System V shared memory), because it allows us to work with names of files in the file system and do away with the keys that are needed in System V.
 
-1. Implement the logger into the wrapper for error handling.
-2. We may potentially need more functionality to support the Gamepad device, and to store robot state (running mode, E-stop, etc.)
-3. More tests could always be used; expanding on the existing testers is probably a good idea.
+The way POSIX shared memory works is that one process first creates the shared memory block of a certain size, and with a predefined name (which is a path in the filesystem). This essentially "creates a file" in the filesystem. Then, the creating process maps the file into certain addresses in its own process space, which allows that process to write to that file simply by writing to locations in its own process space. Now, other processes on the same computer can "open" this newly created "file" and map it to their own process spaces, which results in a situation where writing to this special file from any process updates that file for all processes. When a process is terminated, it needs to ensure that its mappings to the shared memory are closed, and the process which created the shared memory file needs to unlink the name of that file from the file itself, so that the operating system can reclaim that space once there are no more references to that file. If this doesn't to happen, any new attemps to bind the same name to a newly created piece of shared memory will fail.
+
+In the runtime shared memory wrapper, the `dev_handler` is responsible for creating and unlinking. For the auxiliary shared memory wrapper, it is the `net_handler` that is responsible.
+
+For more details and a great example of POSIX shared memory, see this page: https://www.softprayog.in/programming/interprocess-communication-using-posix-shared-memory-in-linux
+
+### Additional Details
+
+Shared memory almost always has to be used in conjunction with semaphores, because you want to prevent proceses from reading and writing the same block of memory at the same time. This is certainly the case in the shared memory wrappers. 
```

Comment:
Think it would be good to define what a semaphore is

Reply:
👌